In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('uk_accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [9]:
accident['Index'] = accident['Index'].astype('category')
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Accident Date'] = accident['Accident Date'].astype(str).str.strip()
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [10]:
accident.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [11]:
accident.isnull().sum()

Index                           0
Accident_Severity               0
Accident Date              395672
Latitude                       25
Light_Conditions                0
District Area                   0
Longitude                      26
Number_of_Casualties            0
Number_of_Vehicles              0
Road_Surface_Conditions       726
Road_Type                    4520
Urban_or_Rural_Area            15
Weather_Conditions          14128
Vehicle_Type                    0
dtype: int64

In [22]:
accident['Accident Date'] = accident['Accident Date'].fillna(pd.to_datetime("1900-01-01"))
accident['Latitude']  = accident['Latitude'].fillna(accident['Latitude'].mean())
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean())
accident['Road_Type'] = accident['Road_Type'].astype('category').fillna("Unknown")
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category').fillna("Unknown")
accident['Weather_Conditions']   = accident['Weather_Conditions'].astype('category').fillna("Unknown")
accident['Urban_or_Rural_Area']  = accident['Urban_or_Rural_Area'].astype('category').fillna("Unknown")

In [23]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [24]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['Dayoftheweek'] = accident['Accident Date'].dt.dayofweek

In [26]:
accident.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
Year                                int32
Month                               int32
Day                                 int32
Dayoftheweek                        int32
dtype: object

<h1>UNIVARIATE ANALYSIS</h1>

In [38]:
df_2019 = accident[accident['Year'] == 2019]

In [39]:
print(f"total number of accident in 2019 is {df_2019['Light_Conditions'].value_counts().sum()}")
df_2019['Light_Conditions'].value_counts()

total number of accident in 2019 is 71867


Light_Conditions
Daylight                       52518
Darkness - lights lit          14124
Darkness - no lighting          4242
Darkness - lighting unknown      719
Darkness - lights unlit          264
Name: count, dtype: int64

<h1>QUESTION#1: How many accidents happened each month in 2019?</h1>

In [42]:
df_2019['Month'] = df_2019['Accident Date'].dt.month_name()

In [43]:
monthly_counts = df_2019['Month'].value_counts().reindex([
    'January','February','March','April','May','June',
    'July','August','September','October','November','December'
])

In [53]:
monthly_counts

Month
January      6010
February     6112
March        5975
April        5695
May          6388
June         6187
July         5777
August       5663
September    5935
October      6050
November     5901
December     6174
Name: count, dtype: int64

<h3>INSIGHT#1: The highest accident in month of 2019 is May with 6388 accidents or 8.89%. Followed by June with 6187 accidents or 8.61%. May and June had higher accident counts likely due to increased holiday travel, longer daylight hours, and heavier road usage during the start of summer in uk.</h3>

<h3>INSIGHT#2: The lowest accident in month of 2019 is August with 5663 accidents or 7.88%. Followed by April with 5695 accidents or 7.92%. the lowest accident counts can possibly because of school breaks, fewer work commutes, and reduced traffic during holiday periods.</h3>

<h3>INSIGHT#3: Accident numbers across months were relatively stable (around 5,600–6,300), showing no extreme seasonal spikes, though May and June were above average while April and August were below average.</h3>

<hr>
<h1>QUESTION#2: What is the distribution of Accident Severity (e.g., Fatal, Serious, Slight) in year 2019?</h1>

In [58]:
severity_counts = df_2019['Accident_Severity'].value_counts()

In [56]:
severity_counts

Accident_Severity
Slight     60928
Serious     9823
Fatal       1116
Name: count, dtype: int64

In [55]:
severity_percent = df_2019['Accident_Severity'].value_counts(normalize=True) * 100

In [57]:
severity_percent

Accident_Severity
Slight     84.778828
Serious    13.668304
Fatal       1.552868
Name: proportion, dtype: float64

<h3>INSIGHT#4: In year 2019 most of the accident severity is minor or slight with a number of 60928 accidents, which has a greater gap between serious with a 9823 accidents and fatal with 1116 accidents. The percetage of slight is almost 85% being the highest accident severity</h3>

<hr>
<h1>QUESTION#3: What is the distribution of Accident Severity (e.g., Fatal, Serious, Slight) in year 2019 for each month?</h1>

In [61]:
severity_monthly = df_2019.groupby(['Month', 'Accident_Severity']).size().unstack().fillna(0)
month_order = ['January','February','March','April','May','June',
               'July','August','September','October','November','December']

In [62]:
severity_monthly = severity_monthly.reindex(month_order)
severity_monthly

Accident_Severity  Fatal  Serious  Slight
Month                                    
January              102      848    5060
February             103      827    5182
March                102      806    5067
April                 94      811    4790
May                   87      904    5397
June                  84      834    5269
July                  92      760    4925
August                83      755    4825
September             90      763    5082
October               90      872    5088
November              91      779    5031
December              98      864    5212

<h3>INSIGHT#5: Slight Accident are higher in each month especially in May and June. while fatal accidents are higher during January to March. So even though the higher accidents is month of May and June they are just minor or moderate accident.</h3>

<hr>
<h1>QUESTION#4: Which district/area had the most accidents in 2019?</h1>

In [66]:
district_counts = df_2019['District Area'].value_counts()
district_counts
print("Top 5 Districts with most accidents in 2019:")
print(district_counts.head())

Top 5 Districts with most accidents in 2019:
District Area
Birmingham      1554
Leeds            950
Glasgow City     721
Manchester       684
Liverpool        628
Name: count, dtype: int64


<h3>INSIGHT#5: Most accidents happened in district or area of Birmingham with 1554 accident in year 2019, there's a high gap between Birmingham and other district.</h3>

<hr>
<h1>QUESTION#5: Which Road Type has the highest number of accidents?</h1>

In [73]:
road_counts = df_2019['Road_Type'].value_counts()
road_counts

Road_Type
Single carriageway    53310
Dual carriageway      11022
Roundabout             4603
One way street         1550
Slip road               775
Unknown                 607
Name: count, dtype: int64

In [72]:
road_percent = df_2019['Road_Type'].value_counts(normalize=True) * 100
road_percent

Road_Type
Single carriageway    74.178691
Dual carriageway      15.336664
Roundabout             6.404887
One way street         2.156762
Slip road              1.078381
Unknown                0.844616
Name: proportion, dtype: float64

<h3>INSIGHT#6: Single Carriageway has the highest number of accidents with 53310 or 74.18%. followed by dual carriageways with 11022 or 15.34%, showing that two-lane and multi-lane roads is a total of nearly 90% of all cases, while accidents on roundabouts, one-way streets, and slip roads were much less frequent.</h3>

<hr>

In [74]:
df_2020 = accident[accident['Year'] == 2020]

<h1>QUESTION#6: What is the daily/weekly trend of accidents in 2020?</h1>

In [94]:
daily_trend = df_2020.groupby(df_2020['Accident Date'].dt.date).size()
daily_trend.head()

Accident Date
2020-01-01    257
2020-01-02    532
2020-01-03    522
2020-01-04    485
2020-01-05    480
dtype: int64

In [96]:
highest_day = daily_trend.idxmax(), daily_trend.max()
highest_day

(datetime.date(2020, 5, 9), np.int64(648))

In [97]:
lowest_day  = daily_trend.idxmin(), daily_trend.min()
lowest_day

(datetime.date(2020, 1, 1), np.int64(257))

<h3>INSIGHT#7: In 2020, daily accident trends showed lower counts on holidays which is january 1, 2020.while weekdays consistently recorded higher accidents, averaging over 500 cases per day.</h3>

In [79]:
weekly_trend = df_2020.groupby(df_2020['Accident Date'].dt.day_name()).size().sort_values(ascending=False)
weekly_trend

Accident Date
Wednesday    10812
Saturday     10809
Tuesday      10114
Friday        9932
Thursday      9773
Sunday        9772
Monday        8951
dtype: int64

<h3>INSIGHT#8: In 2020, weekly accident trends showed lower counts on Monday,while Wednesday and Saturday has the higher counts, showing that midweek and weekend traffic contribute heavily to accidents.</h3>

<hr>
<h1>QUESTIO#7: Did the percentage of slight vs. serious vs. fatal accidents change compared to 2019?</h1>

In [98]:
severity_2019 = accident[accident['Year']==2019]['Accident_Severity'].value_counts(normalize=True) * 100
severity_2020 = accident[accident['Year']==2020]['Accident_Severity'].value_counts(normalize=True) * 100
compare = pd.DataFrame({"2019": severity_2019, "2020": severity_2020, "Change (%)": severity_2020 - severity_2019})
compare

2019       2020  Change (%)
Accident_Severity                                  
Slight             84.778828  85.337001    0.558173
Serious            13.668304  13.320411   -0.347893
Fatal               1.552868   1.342588   -0.210281

<h3>INSIGHT#9: The slight accidents increased slightly from 84.78% in 2019 to 85.34% in 2020 (+0.56%), showing most accidents remained minor.</h3>

<h3>INSIGHT#10: The Serious accidents remain consistent with 13% only decreased in 2020 with -0.35%, showing that there's no changes in accident severity. Same with fatal accidents which remain with 1% only decreased in 2020 with -0.21%</h3>

<HR>
<h1>QUESTION#8: What is the distribution of accidents by Weather Conditions in year 2020?</h1>

In [102]:
weather_2020 = df_2020['Weather_Conditions'].value_counts(normalize = True) * 100
weather_2020

Weather_Conditions
Fine no high winds       75.291820
Raining no high winds    14.432108
Other                     2.924618
Unknown                   2.310335
Raining + high winds      2.011031
Fine + high winds         1.763037
Snowing no high winds     0.655616
Fog or mist               0.466058
Snowing + high winds      0.145376
Name: proportion, dtype: float64

<h3>INSIGHT#11: Most accidents which is 75.3% happened in fine weather without high winds, showing that good weather doesn’t guarantee road safety.</h3>

<h3>INSIGHT#12: Rainy conditions with and without high winds contributed to a total of 16.4% of accidents, highlighting the impact of wet roads on crash risk.</h3>

<h3>INSIGHT#13: Extreme weather such as snow 0.8% and fog 0.5% played only a minor role, meaning accidents occur more often in normal conditions than in severe weather.</h3>

<HR>
<h1>QUESTION#9: Is there a correlation between the number of vehicles involved and the number of casualties in 2020</h1>

In [103]:
vehicles_casualties_corr = df_2020['Number_of_Vehicles'].corr(df_2020['Number_of_Casualties'])
vehicles_casualties_corr

np.float64(0.23378095509947672)

<h3>INSIGHT#14: The correlation is positive but weak, there's a slight correlation between number of vehicles and number of casualties. But there are other more reason for the number of casualties.</h3>

<h1>QUESTION#10: Is there a correlation between Weather Conditions and Accident Severity in 2020?</h1>

In [115]:
from scipy.stats import chi2_contingency

In [121]:
contingency = pd.crosstab(df_2020['Weather_Conditions'], df_2020['Accident_Severity'])
f_oneway, p_value, dof, expected = chi2_contingency(contingency)
chi2

np.float64(126.65753938694847)

In [120]:
p_value

np.float64(2.853587670921335e-19)

<h3>INSIGHT#15: There is a significant correlation between weather conditions and accident severity in 2020, rain, snow or fog can cause serious or fatal accidents.</h3>

<HR></HR>
<h1>QUESTION#11: Which vehicle type was most frequently involved in accidents in 2021?</h1>

In [124]:
df_2021 = accident[accident['Year'] == 2021]

In [128]:
vehicle_acc = df_2021['Vehicle_Type'].value_counts(normalize=True) * 100
vehicle_acc

Vehicle_Type
Car                                      78.197727
Van / Goods 3.5 tonnes mgw or under       5.076165
Motorcycle over 500cc                     3.688871
Bus or coach (17 or more pass seats)      2.543372
Goods 7.5 tonnes mgw and over             2.232062
Motorcycle 125cc and under                2.053739
Taxi/Private hire car                     1.668379
Motorcycle 50cc and under                 1.184791
Motorcycle over 125cc and up to 500cc     1.113764
Other vehicle                             0.873481
Goods over 3.5t. and under 7.5t           0.817566
Agricultural vehicle                      0.273530
Minibus (8 - 16 passenger seats)          0.243305
Pedal cycle                               0.030224
Ridden horse                              0.003022
Data missing or out of range              0.000000
Name: proportion, dtype: float64

<h3>INSIGHT#16: Car are most frequently involved in accidents by the year of 2021 with over 51745 or 78.2%, there's a huge gap in car and other vehicle type. </h3>